### ISSUE TITLE
extend or implements from a generic type

### ISSUE THREAD

**Issue by AdamRose66** [dart-lang/language#3552]
I'd like to be able to do something like this:

`class A<T> extends T { ... }`

or 

`class B<T> implements T { ... }`

**Comment by Wdestroier** [comment]
How are you going to use the B class after that?

**Comment by mateusfccp** [comment]
I am not sure what you are trying to achieve.

What would be the use-case of this esoteric approach? What problem are you trying to solve with this?

**Comment by lrhn** [comment]
That's *very unlikely* to happen.

The problem is that `class A<T> extends T`
* Doesn't know its superclass.
* Therefore cannot invoke any `super.` members.
* Doesn't know its superclass's *constructors* (which its own constructors need to invoke).
* Doesn't know its superclass's *interface*.
* Therefore doesn't know it's *own* interface.
* Can't safely declare *any* members, because they might conflict with the superclass member signatures, which it doesn't know.

Basically, it's impossible to do anything safely or soundly, which means that this particular design is unsuited for the Dart language.

A "class" which doesn't know its *concrete* superclass, but may have an idea about its interface, is what a `mixin` is.
It's a class which takes its superclass as an "argument", but it only gets the superclass at *mixin application*, where it creates a new class with a single fixed interface.

**Comment by luizbarboza** [comment]
What I think he's really trying to say is that he'd like to be able to guarantee that type T is a supertype of A.

Something like: 
```dart
class A<T super A> {}
```

Related to #1674.

**Comment by AdamRose66** [comment]
`class A<T> extends T { ... }`

or

class B<T> implements T { ... }

**Comment by AdamRose66** [comment]
The actual requirement is more for 

`class A<T> implements T` 

than 

`class A<T> extends T`

Or more precisely 

`class A<IF> implements IF`

where A is an abstract interface class.

eg 

```
class Proxy<IF> implements IF
{
   ...
   noSuchMethod( Invocation invocation )
   {
     reflect( port_if ).delegate( invocation );
  }

  IF get port_if => m_if;

  IF m_if;
}

abstract interface class MyInterface
{
  void my_message( int n );
}

Proxy<MyInterface> a = A( ... );

a.my_message( 3 );

```

The idea is that A<IF> is a proxy for an abstract interface that we can use to specify the relationships between interfaces

At the moment I have to do:

```
class Proxy<IF> // no implements 
{
   ...
   noSuchMethod( Invocation invocation )
   {
     reflect( port_if ).delegate( invocation );
  }

  IF get port_if => m_if;

  IF m_if;
}

abstract interface class MyInterface
{
  void my_message( int n );
}

// tedious typing needed here ...
class MyProxy extends Proxy<MyInterface> implements MyInterface
{
}

MyProxy a = A( ... );

a.my_message( 3 );

```

**Comment by AdamRose66** [comment]
Now, I completely agree that using noSuchMethod in this way means that we are bypassing the static checking that would otherwise be a very good thing.

If there were some checkable way of delegating an implementation to an interface, then that would be better.

**Comment by AdamRose66** [comment]
Also, I can hear the chorus of replies "that's what all the meta-programming stuff is for". But waiting for that seems overkill for something that is almost already there.

**Comment by AdamRose66** [comment]
And of course we can ( from the original example ) do:

```
Proxy<MyInterface> a = A( ... );

a.port_if.my_message( 3 );

```

In which case, you can re-interpret my request as the ability to overload the '.' operator :).

**Comment by AdamRose66** [comment]
Having read this through quite a few times, I think I get it.

The reason that 

`class MyProxy extends Proxy<MyInterface> implements MyInterface `

is ok is that the interface is known at this point, and so Dart can do the checking it needs to do to see if this is a well-formed class.

As you might have guessed, I come from a C++ background where 

`template<type IF> class Proxy : public IF
{
...
private:
  IF *m_if;
}; `

is fine ( although we lack the introspection capabilities to do the forwarding to the interface automatically ).

But I assume this is because the template expansion is done during run time in C++ ?

Will the meta-programming stuff allow us to create a macro that creates MyProxy ?
